File: Search in Rotated Sorted Array - Leetcode 33/Search in Rotated Sorted Array - Leetcode 33.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        l = 0
        r = n - 1

        while l < r:
            m = (l + r) // 2

            if nums[m] > nums[r]:
                l = m + 1
            else:
                r = m

        min_i = l

        if min_i == 0:
            l, r = 0, n - 1
        elif target >= nums[0] and target <= nums[min_i - 1]:
            l, r = 0, min_i - 1
        else:
            l, r = min_i, n - 1

        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return m
            elif nums[m] < target:
                l = m + 1
            else:
                r = m - 1
        return -1

# Time Complexity: O(log(n))
# Space Complexity: O(1)

File: Search in Rotated Sorted Array - Leetcode 33/Search in Rotated Sorted Array - Leetcode 33.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int l = 0;
        int r = n - 1;

        while (l < r) {
            int m = (l + r) / 2;

            if (nums[m] > nums[r]) {
                l = m + 1;
            } else {
                r = m;
            }
        }

        int minIndex = l;

        if (minIndex == 0) {
            l = 0;
            r = n - 1;
        } else if (target >= nums[0] && target <= nums[minIndex - 1]) {
            l = 0;
            r = minIndex - 1;
        } else {
            l = minIndex;
            r = n - 1;
        }

        while (l <= r) {
            int m = (l + r) / 2;
            if (nums[m] == target) {
                return m;
            } else if (nums[m] < target) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }

        return -1;
    }
};
